subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1087,6 +1093,29 @@ public void preventFromFurtherMutation() {
             configurationAttributes = new ImmutableAttributeContainerWithErrorMessage(delegatee, this.displayName);
             outgoing.preventFromFurtherMutation();
             canBeMutated = false;
+
+            if (isCanBeConsumed()) {
+                ensureUniqueAttributesOnConsumableConfigurations();
+            }
```

Comment:
Please double check me on this! If this can be called from multiple threads, would `canBeMutated` require synchronization (such as volatile)? A race condition is not possible, but it is possible that changes to this variable from one thread are not visible to another thread without it. Two configurations could both think that the other is still mutable.

Reply:
I do not think this thread safety check is necessary here - there would be other problems with the fields of this class if it was used in this way.

Reply:
I wish it were possible to model this as a test, but I don't know how. The new code might be the first time for which the behavior of the method now explicitly depends on the internal state of another configuration that can change.

I am certain that, if `preventFromFurtherMutation` can be called from more than one thread, that what I've described could happen in theory. It might not happen in practice. This is also all moot if it is never called from multiple threads.

Reply:
I was rereading my comments in the light of a new day and feel they are hard to understand, so I'll try to be more specific.

If two configurations `A` and `B` are duplicates, this is only detected after the second one is made immutable. Without loss of generality, suppose that the second one is `B`.

As the detection code runs in some thread T<sub>b</sub>, it must see that `A` is immutable and see the latest attributes of `A` after it was made immutable. If `A` was made immutable in some other thread T<sub>a</sub>, there is no [happened-before](https://en.wikipedia.org/wiki/Happened-before) relationship between these two sets of instructions, so no guarantee exists that the detection running in T<sub>b</sub> sees those changes that were made in T<sub>a</sub>.

Reply:
I track what you're saying, and agree this would be a problem if multiple threads are locking the same configuration instances.  I'll be sure to check with my more knowledgeable colleagues before merging this work, but I just can't see how that would be the case without similar problems happening all over this class.  The synchronization which DOES exist is just not very thorough.